### PR TITLE
Login Backend Integration

### PR DESCRIPTION
**Pull request by Marcosnr97**
### Summary
Added login integration with backed
Added Interceptor to save access-token, client and uid
Added error handling for invalid credentials with toast message

### CODE REVIEW

File: app/src/main/java/ar/com/wolox/android/example/model/Login.kt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package ar.com.wolox.android.example.model
+
+data class Login(val email: String, val password: String)
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/model/Login.kt]: We usually order the data classes so every `val` is in its own line

File: app/src/main/java/ar/com/wolox/android/example/model/User.kt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package ar.com.wolox.android.example.model
+
+data class User(val uid: String, val email: String, val name: String, val nickname: String)
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/model/User.kt]: Same here.
Also, check the login's response, you'll find some more fields to store.

File: app/src/main/java/ar/com/wolox/android/example/network/repository/UserRepository.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class UserRepository @Inject constructor(private val retrofitServices: RetrofitServices) {
+
+    private val service: UserService
+        get() = retrofitServices.getService(UserService::class.java)
+
+    suspend fun loginUser(loginData: Login) = withContext(Dispatchers.IO) {
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/network/repository/UserRepository.kt:16]: Is the difference between `async`, `withContext` and `launch` clear?

File: app/src/main/java/ar/com/wolox/android/example/ui/home/HomeFragment.kt [hunk 4]
```diff
@@ -9,6 +9,11 @@ class HomeFragment private constructor() : WolmoFragment<FragmentHomeBinding, Ho
     override fun layout() = R.layout.fragment_home
 
     override fun init() {
+        presenter.onViewAttached()
+    }
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/home/HomeFragment.kt]: You don't need to call `onViewAttached` from here. `onViewAttached` is triggered every time the fragment is instantiated and attached to the presenter.

File: app/src/main/java/ar/com/wolox/android/example/ui/login/LoginFragment.kt [hunk 5]
```diff
@@ -5,10 +5,14 @@ import ar.com.wolox.android.databinding.FragmentLoginBinding
 import ar.com.wolox.android.example.ui.home.HomeActivity
 import ar.com.wolox.android.example.ui.signup.SignupActivity
 import ar.com.wolox.wolmo.core.fragment.WolmoFragment
+import ar.com.wolox.wolmo.core.util.ToastFactory
 import ar.com.wolox.wolmo.core.util.openBrowser
+import javax.inject.Inject
 
 class LoginFragment : WolmoFragment<FragmentLoginBinding, LoginPresenter>(), LoginView {
 
+    @Inject internal lateinit var toastFactory: ToastFactory
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/login/LoginFragment.kt:15]: Do you need to make this `internal`?

File: app/src/main/java/ar/com/wolox/android/example/ui/login/LoginPresenter.kt [hunk 6]
```diff
@@ -1,33 +1,46 @@
...
     }
 
-    fun onLoginButtonClicked(email: String, password: String) {
+    fun onLoginButtonClicked(email: String, password: String) = launch {
         userSession.email = email
         userSession.password = password
-        var loginError = false
+        if (password.isBlank() && email.isBlank()) {
+            view?.showEmptyPasswordAndEmailError()
+            return@launch
```
Comment (ffasitella) [inline app/src/main/java/ar/com/wolox/android/example/ui/login/LoginPresenter.kt]: I think you don't need to return anything

File: app/src/main/res/layout/fragment_home.xml [hunk 7]
```diff
@@ -5,5 +5,13 @@
...
         android:layout_width="match_parent"
         android:layout_height="match_parent">
+        <TextView
+            android:id="@+id/EmailText"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/spacing_medium"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintTop_toTopOf="parent"/>
```
Comment (ffasitella) [inline app/src/main/res/layout/fragment_home.xml]: This is a placeholder view, but don't forget to separate the first and last row in this block with a blank space

**Review by ffasitella — CHANGES_REQUESTED** [review]
Good job, we need to fix some small stuff 😄 
Also, we usually add a short gif or video to every PR, to show how that feature works

File: app/src/main/java/ar/com/wolox/android/example/BootstrapApplication.kt [hunk 8]
```diff
@@ -2,18 +2,24 @@ package ar.com.wolox.android.example
...
 import okhttp3.logging.HttpLoggingInterceptor.Level
+import javax.inject.Inject
 
 class BootstrapApplication : WolmoApplication() {
 
+    @Inject
+    lateinit var userSession: UserSession
+
     override fun onInit() {
         // Initialize Application stuff here
```
Comment (jcruz-wolox) [inline app/src/main/java/ar/com/wolox/android/example/BootstrapApplication.kt]: Delete this comment, isn't need.

File: app/src/main/res/layout/fragment_home.xml (hunk 7 above)
Comment (jcruz-wolox) [reply app/src/main/res/layout/fragment_home.xml]: You can use the Android Studio tool to make code reformat, you have go to Code -> Reformat Code.

File: app/src/main/java/ar/com/wolox/android/example/network/repository/UserRepository.kt (hunk 3 above)
Comment (Marcosnr97) [reply app/src/main/java/ar/com/wolox/android/example/network/repository/UserRepository.kt:16]: :ok_hand:

**Review by ffasitella — APPROVED** [review]
Good job! 👍🏽

**Pull request merged**